# **Graded Challenge 3**

##### **Nama : Achmad Naufal Wijaya J**
##### **Batch : 002**
_Graded Challenge ini dibuat guna mengevaluasi pembelajaran pada Hacktiv8 Data Science Fulltime Program khususnya pada Practical Statistics._

---

## **Dataset Description**

Unduh dataset yang akan digunakan [disini](https://www.kaggle.com/kemical/kickstarter-projects?select=ks-projects-201801.csv).

## **Assignment Instructions**

*Graded Challenge 1* dikerjakan dalam format ***notebook*** dengen beberapa **kriteria wajib** di bawah ini:

1. *Library* yang digunakan adalah **Numpy**, **Pandas**, dan **Scipy**.
2. *Project* dinyatakan selesai dan diterima untuk dinilai jika saat dilakukan `Run All` pada *notebook*, semua *cell* berhasil tereksekusi sampai akhir.
3. Isi *notebook* harus mengikuti *outline* di bawah ini:
   1. Perkenalan\
   Bab pengenalan harus diisi dengan identitas, gambaran besar dataset yang digunakan, dan *objective* yang ingin dicapai.
   2. *Import* pustaka yang dibutuhkan\
   *Cell* pertama pada *notebook* **harus berisi dan hanya berisi** semua *library* yang digunakan dalam *project*.
   3. *Data Loading*\
   Bagian ini berisi proses *data loading* yang kemudian dilanjutkan dengan *explorasi data* secara sederhana.
   4. *Data Cleaning*\
   Bagian ini berisi proses penyiapan data berupa data cleaning sebelum dilakukan *explorasi data* lebih lanjut. Proses cleaning dapat berupa memberi nama baru untuk setiap kolom, mengisi missing values, menghapus kolom yang tidak dipakai, dan lain sebagainya.
   5. *Explorasi Data*\
   Bagian ini berisi explorasi data pada dataset diatas dengan menggunakan query, grouping, visualisasi sederhana, dan lain sebagainya.
   6. *Measure of Central Tendency*
   7. *Measure of Variance*
   8. *Hyphotesis Testing*
   9. Pengambilan Kesimpulan\
   Pada bab terakhir ini, **harus berisi** kesimpulan yang mencerminkan hasil yang didapat dengan dibandingkan dengan *objective* yang sudah ditulis di bagian pengenalan.
4. *Notebook* harus diupload dalam akun GitHub masing-masing siswa untuk selanjutnya dinilai.

## **Assignment Submission**

- Simpan assignment pada sesi ini dengan nama `h8dsft_P0W3`.
- Push Assigment yang telah kalian buat ke akun Github kalian masing-masing.

## **Assignment Objectives**

*Graded Challenge 1* ini dibuat guna mengevaluasi kPractical Statistics sebagai berikut:

- Mampu memuat data dengan Pandas
- Mampu menangani missing values pada dataset
- Mampu melakukan manipulasi kolom
- Mampu membuat query terhadap dataset
- Mampu melakukan grouping terhadap dataset
- Mampu melakukan visualisasi sederhana menggunakan Pandas

---

## **Import Library**

In [27]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
from scipy.stats import t as t_func
from scipy.stats import norm
from scipy.stats import kurtosis

## **Data Loading**

In [3]:
df = pd.read_csv("ks-projects-201801.csv")
df.head()   #Melihat 5 data teratas

ID                                               name  \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
2  1000004038                                     Where is Hank?   
3  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
4  1000011046  Community Film Project: The Art of Neighborhoo...   

         category main_category currency    deadline     goal  \
0          Poetry    Publishing      GBP  2015-10-09   1000.0   
1  Narrative Film  Film & Video      USD  2017-11-01  30000.0   
2  Narrative Film  Film & Video      USD  2013-02-26  45000.0   
3           Music         Music      USD  2012-04-16   5000.0   
4    Film & Video  Film & Video      USD  2015-08-29  19500.0   

              launched  pledged     state  backers country  usd pledged  \
0  2015-08-11 12:12:28      0.0    failed        0      GB          0.0   
1  2017-09-02 04:43:57   2421.0    failed       15      US        100.0   
2  2013-01-12 00:20:50    220.0    failed        3      US        220.0   
3  2012-03-17 03:24:11      1.0    failed        1      US          1.0   
4  2015-07-04 08:35:03   1283.0  canceled       14      US       1283.0   

   usd_pledged_real  usd_goal_real  
0               0.0        1533.95  
1            2421.0       30000.00  
2             220.0       45000.00  
3               1.0        5000.00  
4            1283.0       19500.00

## **Data Cleaning**

Sebelum kita melakukan *cleaning data*, mari kita periksa dulu apakah ada data yang masih kosong / *Null*

In [4]:
df.isnull().sum()  

ID                     0
name                   4
category               0
main_category          0
currency               0
deadline               0
goal                   0
launched               0
pledged                0
state                  0
backers                0
country                0
usd pledged         3797
usd_pledged_real       0
usd_goal_real          0
dtype: int64

Ternyata setelah dilihat, masih terdapat $4$ data pada **name** yang masih kosong dan $3797$ data pada **usd_pledged** yang masih kosong

Karena hal tersebut, mari kita lakukan *cleansing data* dengan cara menghapus data dengan **name** yang kosong dan mengganti data-data pada **usd pledged** dengan nilai *median*

In [5]:
df = df[df['name'].isna() == False]
df.head(5)

ID                                               name  \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
2  1000004038                                     Where is Hank?   
3  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
4  1000011046  Community Film Project: The Art of Neighborhoo...   

         category main_category currency    deadline     goal  \
0          Poetry    Publishing      GBP  2015-10-09   1000.0   
1  Narrative Film  Film & Video      USD  2017-11-01  30000.0   
2  Narrative Film  Film & Video      USD  2013-02-26  45000.0   
3           Music         Music      USD  2012-04-16   5000.0   
4    Film & Video  Film & Video      USD  2015-08-29  19500.0   

              launched  pledged     state  backers country  usd pledged  \
0  2015-08-11 12:12:28      0.0    failed        0      GB          0.0   
1  2017-09-02 04:43:57   2421.0    failed       15      US        100.0   
2  2013-01-12 00:20:50    220.0    failed        3      US        220.0   
3  2012-03-17 03:24:11      1.0    failed        1      US          1.0   
4  2015-07-04 08:35:03   1283.0  canceled       14      US       1283.0   

   usd_pledged_real  usd_goal_real  
0               0.0        1533.95  
1            2421.0       30000.00  
2             220.0       45000.00  
3               1.0        5000.00  
4            1283.0       19500.00

In [6]:
df['usd pledged'].fillna(df['usd pledged'].median(), inplace = True)
df.head(5)

ID                                               name  \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
2  1000004038                                     Where is Hank?   
3  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
4  1000011046  Community Film Project: The Art of Neighborhoo...   

         category main_category currency    deadline     goal  \
0          Poetry    Publishing      GBP  2015-10-09   1000.0   
1  Narrative Film  Film & Video      USD  2017-11-01  30000.0   
2  Narrative Film  Film & Video      USD  2013-02-26  45000.0   
3           Music         Music      USD  2012-04-16   5000.0   
4    Film & Video  Film & Video      USD  2015-08-29  19500.0   

              launched  pledged     state  backers country  usd pledged  \
0  2015-08-11 12:12:28      0.0    failed        0      GB          0.0   
1  2017-09-02 04:43:57   2421.0    failed       15      US        100.0   
2  2013-01-12 00:20:50    220.0    failed        3      US        220.0   
3  2012-03-17 03:24:11      1.0    failed        1      US          1.0   
4  2015-07-04 08:35:03   1283.0  canceled       14      US       1283.0   

   usd_pledged_real  usd_goal_real  
0               0.0        1533.95  
1            2421.0       30000.00  
2             220.0       45000.00  
3               1.0        5000.00  
4            1283.0       19500.00

In [7]:
df.isnull().sum()

ID                  0
name                0
category            0
main_category       0
currency            0
deadline            0
goal                0
launched            0
pledged             0
state               0
backers             0
country             0
usd pledged         0
usd_pledged_real    0
usd_goal_real       0
dtype: int64

Karena data sudah tidak ada yang kosong lagi, selanjutnya kita bisa melakukan ke tahap **explorasi data**

## **Data Exploration**

In [8]:
df.head(5)

ID                                               name  \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
2  1000004038                                     Where is Hank?   
3  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
4  1000011046  Community Film Project: The Art of Neighborhoo...   

         category main_category currency    deadline     goal  \
0          Poetry    Publishing      GBP  2015-10-09   1000.0   
1  Narrative Film  Film & Video      USD  2017-11-01  30000.0   
2  Narrative Film  Film & Video      USD  2013-02-26  45000.0   
3           Music         Music      USD  2012-04-16   5000.0   
4    Film & Video  Film & Video      USD  2015-08-29  19500.0   

              launched  pledged     state  backers country  usd pledged  \
0  2015-08-11 12:12:28      0.0    failed        0      GB          0.0   
1  2017-09-02 04:43:57   2421.0    failed       15      US        100.0   
2  2013-01-12 00:20:50    220.0    failed        3      US        220.0   
3  2012-03-17 03:24:11      1.0    failed        1      US          1.0   
4  2015-07-04 08:35:03   1283.0  canceled       14      US       1283.0   

   usd_pledged_real  usd_goal_real  
0               0.0        1533.95  
1            2421.0       30000.00  
2             220.0       45000.00  
3               1.0        5000.00  
4            1283.0       19500.00

In [9]:
df.tail(5)

ID                                               name  \
378656  999976400  ChknTruk Nationwide Charity Drive 2014 (Canceled)   
378657  999977640                                          The Tribe   
378658  999986353  Walls of Remedy- New lesbian Romantic Comedy f...   
378659  999987933                           BioDefense Education Kit   
378660  999988282                  Nou Renmen Ayiti!  We Love Haiti!   

               category main_category currency    deadline     goal  \
378656      Documentary  Film & Video      USD  2014-10-17  50000.0   
378657   Narrative Film  Film & Video      USD  2011-07-19   1500.0   
378658   Narrative Film  Film & Video      USD  2010-08-16  15000.0   
378659       Technology    Technology      USD  2016-02-13  15000.0   
378660  Performance Art           Art      USD  2011-08-16   2000.0   

                   launched  pledged     state  backers country  usd pledged  \
378656  2014-09-17 02:35:30     25.0  canceled        1      US         25.0   
378657  2011-06-22 03:35:14    155.0    failed        5      US        155.0   
378658  2010-07-01 19:40:30     20.0    failed        1      US         20.0   
378659  2016-01-13 18:13:53    200.0    failed        6      US        200.0   
378660  2011-07-19 09:07:47    524.0    failed       17      US        524.0   

        usd_pledged_real  usd_goal_real  
378656              25.0        50000.0  
378657             155.0         1500.0  
378658              20.0        15000.0  
378659             200.0        15000.0  
378660             524.0         2000.0

In [10]:
df.info()   #Melihat tipe data

<class 'pandas.core.frame.DataFrame'>
Int64Index: 378657 entries, 0 to 378660
Data columns (total 15 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   ID                378657 non-null  int64  
 1   name              378657 non-null  object 
 2   category          378657 non-null  object 
 3   main_category     378657 non-null  object 
 4   currency          378657 non-null  object 
 5   deadline          378657 non-null  object 
 6   goal              378657 non-null  float64
 7   launched          378657 non-null  object 
 8   pledged           378657 non-null  float64
 9   state             378657 non-null  object 
 10  backers           378657 non-null  int64  
 11  country           378657 non-null  object 
 12  usd pledged       378657 non-null  float64
 13  usd_pledged_real  378657 non-null  float64
 14  usd_goal_real     378657 non-null  float64
dtypes: float64(5), int64(2), object(8)
memory usage: 46.2+ MB


In [11]:
df.shape  #Melihat ukuran data

(378657, 15)

In [12]:
df.describe()   #Melihat statistika secara numerik

ID          goal       pledged        backers   usd pledged  \
count  3.786570e+05  3.786570e+05  3.786570e+05  378657.000000  3.786570e+05   
mean   1.074732e+09  4.908068e+04  9.683080e+03     105.618544  6.970199e+03   
std    6.190872e+08  1.183397e+06  9.563651e+04     907.189767  7.824768e+04   
min    5.971000e+03  1.000000e-02  0.000000e+00       0.000000  0.000000e+00   
25%    5.382499e+08  2.000000e+03  3.000000e+01       2.000000  1.900000e+01   
50%    1.075276e+09  5.200000e+03  6.200000e+02      12.000000  3.947650e+02   
75%    1.610149e+09  1.600000e+04  4.076000e+03      56.000000  3.000000e+03   
max    2.147476e+09  1.000000e+08  2.033899e+07  219382.000000  2.033899e+07   

       usd_pledged_real  usd_goal_real  
count      3.786570e+05   3.786570e+05  
mean       9.059018e+03   4.545425e+04  
std        9.097382e+04   1.152956e+06  
min        0.000000e+00   1.000000e-02  
25%        3.100000e+01   2.000000e+03  
50%        6.243800e+02   5.500000e+03  
75%        4.050000e+03   1.550000e+04  
max        2.033899e+07   1.663614e+08

## **Measure of Central Tendency**

***Measure of central tendency*** adalah statistik ringkasan yang mewakili titik pusat atau nilai khas dari kumpulan data. Ada tiga ukuran utama tendensi sentral:

- **Mean** : jumlah nilai setiap pengamatan dalam kumpulan data dibagi dengan jumlah pengamatan

- **Mode** : nilai yang paling sering muncul dalam suatu distribusi

- **Median** : nilai tengah dalam distribusi ketika nilai-nilai diatur dalam urutan naik atau turun

In [13]:
# Mean
df[df.describe().columns].mean(axis = 0)

ID                  1.074732e+09
goal                4.908068e+04
pledged             9.683080e+03
backers             1.056185e+02
usd pledged         6.970199e+03
usd_pledged_real    9.059018e+03
usd_goal_real       4.545425e+04
dtype: float64

In [14]:
# Modus
list = df.columns.values
df[list].mode(axis = 0, numeric_only = True)

ID    goal  pledged  backers  usd pledged  usd_pledged_real  \
0             5971  5000.0      0.0      0.0          0.0               0.0   
1            18520     NaN      NaN      NaN          NaN               NaN   
2            21109     NaN      NaN      NaN          NaN               NaN   
3            21371     NaN      NaN      NaN          NaN               NaN   
4            24380     NaN      NaN      NaN          NaN               NaN   
...            ...     ...      ...      ...          ...               ...   
378652  2147455254     NaN      NaN      NaN          NaN               NaN   
378653  2147460119     NaN      NaN      NaN          NaN               NaN   
378654  2147466649     NaN      NaN      NaN          NaN               NaN   
378655  2147472329     NaN      NaN      NaN          NaN               NaN   
378656  2147476221     NaN      NaN      NaN          NaN               NaN   

        usd_goal_real  
0              5000.0  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
...               ...  
378652            NaN  
378653            NaN  
378654            NaN  
378655            NaN  
378656            NaN  

[378657 rows x 7 columns]

In [15]:
# Median
list = df.columns.values
df[list].median()

ID                  1.075276e+09
goal                5.200000e+03
pledged             6.200000e+02
backers             1.200000e+01
usd pledged         3.947650e+02
usd_pledged_real    6.243800e+02
usd_goal_real       5.500000e+03
dtype: float64

## **Measure of Variance**

Measure of variability is ringkasan statistik yang mewakili jumlah dispersi dalam kumpulan data. Ada 4 pengukuran variabilitas yang paling umum digunakan :

- Range: perbedaan antara nilai tertinggi dan terendah dari kumpulan data

- Variance: perbedaan kuadrat rata-rata dari nilai-nilai dari mean

- Standard Deviation : perbedaan standar atau tipikal antara setiap titik data dan rata-rata

- Interquartile Range : kisaran setengah tengah distribusi

In [16]:
# Range
list = df.describe().columns
print(df[list].max()-df[list].min())

ID                  2.147470e+09
goal                1.000000e+08
pledged             2.033899e+07
backers             2.193820e+05
usd pledged         2.033899e+07
usd_pledged_real    2.033899e+07
usd_goal_real       1.663614e+08
dtype: float64


In [17]:
# Variance
df.var()

ID                  3.832690e+17
goal                1.400430e+12
pledged             9.146342e+09
backers             8.229933e+05
usd pledged         6.122700e+09
usd_pledged_real    8.276236e+09
usd_goal_real       1.329308e+12
dtype: float64

In [18]:
# Standard Deviation
df.std()

ID                  6.190872e+08
goal                1.183397e+06
pledged             9.563651e+04
backers             9.071898e+02
usd pledged         7.824768e+04
usd_pledged_real    9.097382e+04
usd_goal_real       1.152956e+06
dtype: float64

In [19]:
#Interquartile Range (IQR)
IQR = df[list].quantile(0.75) - df[list].quantile(0.25)
IQR

ID                  1.071899e+09
goal                1.400000e+04
pledged             4.046000e+03
backers             5.400000e+01
usd pledged         2.981000e+03
usd_pledged_real    4.019000e+03
usd_goal_real       1.350000e+04
dtype: float64

## **Hypotesis Testing**

Dalam proyek ini, kami akan menguji hubungan antara ***category*** dan ***state*** proyek kickstarter. Jadi, kami mendefinisikan untuk :

- H0 : Tidak ada hubungan antara category dan state

- H1 : Ada hubungan antara category dan state

Dalam uji hipotesis ini kita akan menggunakan uji chi square

In [20]:
df_HT = df.groupby(['category','state']).count()[['goal','pledged']].reset_index()
df_HT.head(10)

category       state  goal  pledged
0  3D Printing    canceled    91       91
1  3D Printing      failed   326      326
2  3D Printing        live     8        8
3  3D Printing  successful   242      242
4  3D Printing   suspended    15       15
5  3D Printing   undefined     1        1
6     Academic    canceled   115      115
7     Academic      failed   589      589
8     Academic        live    11       11
9     Academic  successful   188      188

In [21]:
df_status=df_HT.pivot(index="category",columns="state", values="pledged").fillna(0)
df_status.head(10)

state         canceled  failed   live  successful  suspended  undefined
category                                                               
3D Printing       91.0   326.0    8.0       242.0       15.0        1.0
Academic         115.0   589.0   11.0       188.0       13.0        0.0
Accessories      340.0  1667.0   53.0      1073.0       29.0        3.0
Action           109.0   514.0    7.0       107.0        3.0        0.0
Animals           18.0   166.0    3.0        63.0        5.0        0.0
Animation        306.0  1531.0   16.0       682.0        6.0        0.0
Anthologies       37.0   216.0    9.0       521.0        1.0        0.0
Apparel          864.0  4752.0  109.0      1414.0       27.0        0.0
Apps             952.0  4909.0   79.0       378.0       27.0        0.0
Architecture      96.0   463.0   11.0       179.0       11.0        0.0

In [22]:
chisq, pvalue, df, expected = stats.chi2_contingency(df_status)
print(f'Observed chi2: {chisq:.2f}')
print(f'p-value: {pvalue:.4f}')

Observed chi2: 83380.23
p-value: 0.0000


Dari uji chi square dapat diketahui bahwa nilai chi sebesar **5300.05** dengan nilai $p$ sebesar 0.0000 

Karena nilai $p$ lebih kecil dari $0.05$ maka **null hypotesis** ditolak dan disimpulkan bahwa terdapat hubungan antara variabel **category** dan variabel **state**

## **Conclusion**

Dari uji chi square dapat diketahui bahwa nilai chi sebesar **5300.05** dengan nilai $p$ sebesar 0.0000 

Karena nilai $p$ lebih kecil dari $0.05$ maka **null hypotesis** ditolak dan disimpulkan bahwa terdapat hubungan antara variabel **category** dan variabel **state**

**Measure of Central Tendency**
- **Mean** : jumlah nilai setiap pengamatan dalam kumpulan data dibagi dengan jumlah pengamatan. Pencarian mean dapat menggunakan fungsi **.mean()**

- **Mode** : nilai yang paling sering muncul dalam suatu distribusi. Pencarian mode dapat menggunakan fungsi **.mode()**

- **Median** : nilai tengah dalam distribusi ketika nilai-nilai diatur dalam urutan naik atau turun. Pencarian median dapat menggunakan fungsi **.median()**

**Measure of Variance**
- **Range**: perbedaan antara nilai tertinggi dan terendah dari kumpulan data. Pencarian range dapat menggunakan selisih dari fungsi **.min()** dan **.max()**

- **Variance**: perbedaan kuadrat rata-rata dari nilai-nilai dari mean. Pencarian variance dapat menggunakan fungsi **.var()** 

- **Standard Deviation** : perbedaan standar atau tipikal antara setiap titik data dan rata-rata. Pencarian standard deviation dapat menggunakan fungsi **.std()** 

- **Interquartile Range** : kisaran setengah tengah distribusi. Pencarian interquartile range dapat menggunakan selisih dari fungsi **.quantile(0.75)** untuk $Q3$ dan **.quantile(0.25)** untuk $Q1$